### PR TITLE
8346388: Cannot use DllMain in libawt for static builds

### PR DESCRIPTION
**Pull request by magicus**
To be able to properly support static builds on Windows in [JDK-8346377](https://bugs.openjdk.org/browse/JDK-8346377), we cannot use `DllMain`, for two reasons:

1) This is not called for statically linked libraries, and
2) There are multiple `DllMain` definitions throughout the JDK native libraries, causing name collisions.

While it could have been possible to keep the `DllMain` function for non-static builds and just use an alternative solution for static builds, I think it is preferable to have a single solution that works as well for both static and dynamic builds.

In this case, the `DllMain` function did two things: 

1) At startup, it called SetModuleHandle. This has been moved to the `JNI_OnLoad` function, which is called by the JVM right after loading the DLL, or in the case of a static build, when the `awt` native library has been requested.

2) At DLL unload -- for debug builds only -- it "disabled" the mutexes for the DTrace and DMem debug systems. In this case, "disable" means writing NULL to the mutexes, causing any further calls to the continue without locking, since the enter/exit calls only do the locking if the mutex is not NULL. (This is pre-existing code so I am not discussing the soundness of this approach.) 

But why did we need to do that? After the DLL unloading, which is done by Windows when the process is exiting, no code should be executing in AWT, right? No, wrong. There are three static objects, one instance of AwtDebugSupport and two instances of GDIHashtable, and the destructors of these objects are called by Windows at the time of process shutdown. I have not been able to confirm that the `DllMain` code is guaranteed to be called before the destructors are called, so I guess the fact that this ever worked has just been a lucky coincidence. 

I have solved this by disabling the mutexes in the destructors themselves, thereby guaranteeing that they are disabled before the last few calls to the DTrace/DMem calls are made. There is no guarantee in which order these destructors are made, so I do the same on both locations. (There is no bad effect from calling these twice in a row, it's just setting a mutex to NULL).

Finally, I have wrapped the `DEF_JNI_OnLoad` function in `extern "C" { ... }`. This is needed since DEF_JNI_OnLoad is a macro that actually creates two functions, where one is a wrapper calling the other which has a generated name, and without this, the C++ name mangling messed things up. (To be clear: this means that the JNI_OnLoad function has never been called on Windows, but since it did not do anything of substance before, nobody noticed this.)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8346388](https://bugs.openjdk.org/browse/JDK-8346388): Cannot use DllMain in libawt for static builds (**Bug** - P3)

### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) Review applies to [961b31ca](https://git.openjdk.org/jdk/pull/22790/files/961b31ca80e9ccdff4d96184e3b5c9284ad725ec)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22790/head:pull/22790` \
`$ git checkout pull/22790`

Update a local copy of the PR: \
`$ git checkout pull/22790` \
`$ git pull https://git.openjdk.org/jdk.git pull/22790/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22790`

View PR using the GUI difftool: \
`$ git pr show -t 22790`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22790.diff">https://git.openjdk.org/jdk/pull/22790.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22790#issuecomment-2549242942)
</details>

### CODE REVIEW

**Comment by magicus** [comment]
I would appreciate some suggestions on how to make sure this is properly tested, with both release and debug build.

File: src/java.desktop/windows/native/libawt/windows/awt_Toolkit.cpp [hunk 1]
```diff
@@ -102,17 +102,36 @@ extern void DWMResetCompositionEnabled();
...
 JavaVM *jvm = NULL;
 
+/* Return a handle to the module containing this method, either a DLL in case
+ * of a dynamic library build, or the .EXE in case of a static build.
+ */
+HMODULE GetAwtModuleHandle() {
+    HMODULE hModule = NULL;
+    GetModuleHandleEx(
+        GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
+        (LPCTSTR) &GetAwtModuleHandle,
```
Comment (TheShermanTanker) [inline src/java.desktop/windows/native/libawt/windows/awt_Toolkit.cpp:112]: I had to look up the docs on why this was casting a function pointer to a string because it was just so confusing. Wow that threw me for a loop

**Comment by TheShermanTanker** [comment]
AWT tests are in tier 4, to my knowledge

**Comment by prrace** [comment]
>  There are multiple DllMain definitions throughout the JDK native libraries, causing name collisions.

DllMain is a default name, not a required name.
You could just rename them and add /entry to the linker command line, eg /entry:AwtDllMain

https://learn.microsoft.com/en-us/cpp/build/reference/entry-entry-point-symbol?view=msvc-170&viewFallbackFrom=vs-2019

I'm not sure it seems right to have every static destructor need to know what needs to be done to exit the DLL.
I do read your comment about not knowing when the static destructors are called vs the DllMain on shutdown, but surely this *is* well defined ?

**Comment by magicus** [comment]
> DllMain is a default name, not a required name. You could just rename them and add /entry to the linker command line, eg /entry:AwtDllMain

That was an interesting piece of information I did not know. That could be used to resolve issue 2. However, it does not resolve issue 1, which is that this code must be called somehow for static libraries nevertheless.

> I do read your comment about not knowing when the static destructors are called vs the DllMain on shutdown, but surely this _is_ well defined ?

That is possible. It could be just that I have failed to locate the documentation about this. I did find about the static initializer fiasco, and I have studied when JNI_OnUnload happens (answer: basically it doesn't), and what I have found about DllMain has not been very specific, so I'm not too hopeful that it is indeed well defined.

**Comment by magicus** [comment]
> I'm not sure it seems right to have every static destructor need to know what needs to be done to exit the DLL.

That is an over-generalization of the problem at hand that is not really true. What we have is an unfortunate coupling between the API from one DLL (JVM.DLL) and another (AWT.DLL), and that only happens in three specific static objects, each of which are very specialized.

Another way of describing the problem is that the AWT debug system should not use JVM raw monitors, since they could happen to be used when the JVM is unloaded and the calls will fail. In fact, my first solution attempt for this bug was to replace the locking mechanism with an own implementation of pthread mutexes/critical sections. (Basically a simple re-implementation of JVM raw monitors.)

The current code tries to solve this by "disabling" the locks when awt.dll is unloaded, but technically it should be disabled when the jvm.dll module is unloaded -- the fact that this work is just due to the circumstance that they typically are unloaded together when the process ends. I don't know if it would be possible to trigger some kind of race, where jvm.dll is unloaded but some thread is still running the awt debug code, like some of the `operator delete` stuff (which does take a lock).

I think the entire design leaves a lot to be desired, but I thought this was the most elegant to the solution to the problem at hand. As a second hand alternative, I'd suggest replacing JVM raw monitors with a hand-rolled implementation.

**Comment by TheShermanTanker** [comment]
> > DllMain is a default name, not a required name. You could just rename them and add /entry to the linker command line, eg /entry:AwtDllMain
> 
> That was an interesting piece of information I did not know. That could be used to resolve issue 2. However, it does not resolve issue 1, which is that this code must be called somehow for static libraries nevertheless.

That would not be the correct solution to issue 2. Contrary to popular belief DllMain is _not_ the first piece of code that is called for a dll (That is, it is not the actual dll entry point), that would be DllMainCRTStartup (Or something similar, I forgot the exact name). \-entry changes the default name of the initial loader from DllMainCRTStartup to whatever you specified, NOT the DllMain initialization function. Using \-entry in this way would royally break AWT and make it non\-functional. If that doesn't happen, it would mean we are monumentally lucky, but it would likely still not work properly

**Comment by magicus** [comment]
@prrace Are you okay with this solution, or would you prefer that I replace the use of JVM raw monitors with a hand-rolled mutex implementation?

**Comment by magicus** [comment]
Actually, when going over the backlog from my vacation, it looks like JDK-8346306 might fix this problem from the Hotspot side. I'll have to check that, and if so, this PR can be made much smaller.

**Comment by magicus** [comment]
> > I do read your comment about not knowing when the static destructors are called vs the DllMain on shutdown, but surely this _is_ well defined ?
> 
> That is possible. It could be just that I have failed to locate the documentation about this.

Here is the documentation: (from https://learn.microsoft.com/en-us/cpp/build/run-time-library-behavior?view=msvc-170; thanks @dholmes-ora for the link)

"The VCRuntime library provides an entry-point function called _DllMainCRTStartup to handle default initialization and termination operations. On process attach, the _DllMainCRTStartup function sets up buffer security checks, initializes the CRT and other libraries, initializes run-time type information, initializes and calls constructors for static and non-local data, initializes thread-local storage, increments an internal static counter for each attach, and then calls a user- or library-supplied DllMain. On process detach, the function goes through these steps in reverse. It calls DllMain, decrements the internal counter, calls destructors, calls CRT termination functions and registered atexit functions, and notifies any other libraries of termination."

The TL;DR is that `DllMain` with `DLL_PROCESS_DETACH` happens before static destructors.

**Comment by magicus** [comment]
It turns out that JDK-8346306 indeed solves part of this problem -- we don't need to disable the raw JVM monitors, which simplifies this code a bit.

**Comment by magicus** [comment]
@prrace Are you okay with this?

**Review by aivanov-jdk — APPROVED** [review]
Looks good to me.

**Comment by aivanov-jdk** [comment]
> Finally, I have wrapped the `DEF_JNI_OnLoad` function in `extern "C" { ... }`. This is needed since `DEF_JNI_OnLoad` is a macro that actually creates two functions, where one is a wrapper calling the other which has a generated name, and without this, the C++ name mangling messed things up. (To be clear: this means that the `JNI_OnLoad` function has never been called on Windows, but since it did not do anything of substance before, nobody noticed this.)

I assumed that `JNI_OnLoad` would not be exported correctly if it were not declared with `extern "C"` modifier. At the same time, I see `JNI_OnLoad` in the list of exported functions from `awt.dll` in a recent build of JDK.

**Comment by aivanov-jdk** [comment]
> > I have solved this by disabling the mutexes in the destructors themselves, thereby guaranteeing that they are disabled before the last few calls to the DTrace/DMem calls are made. There is no guarantee in which order these destructors are made, so I do the same on both locations. (There is no bad effect from calling these twice in a row, it's just setting a mutex to NULL).
>
> Actually, when going over the backlog from my vacation, it looks like JDK-8346306 might fix this problem from the Hotspot side. I'll have to check that, and if so, this PR can be made much smaller.

[JDK-8346306](https://bugs.openjdk.org/browse/JDK-8346306) is the reason why no destructors are modified anymore, is it?

File: src/java.desktop/windows/native/libawt/windows/awt_Toolkit.cpp [hunk 2]
```diff
@@ -102,17 +102,36 @@ extern void DWMResetCompositionEnabled();
    first loaded */
 JavaVM *jvm = NULL;
 
+/* Return a handle to the module containing this method, either a DLL in case
+ * of a dynamic library build, or the .EXE in case of a static build.
+ */
+HMODULE GetAwtModuleHandle() {
```
[the remark this answers is not in the document]
Comment (magicus) [reply src/java.desktop/windows/native/libawt/windows/awt_Toolkit.cpp:108]: Actually, this function needs to be **not** static for it to work -- it uses it's own address to look up the handle that contains it -- which will be awt.dll when running a normal, dynamic build, and the entire java.exe when running a static build.

**Comment by magicus** [comment]
> I assumed that `JNI_OnLoad` would not be exported correctly if it were not declared with `extern "C"` modifier. At the same time, I see `JNI_OnLoad` in the list of exported functions from `awt.dll` in a recent build of JDK.

Yeah, I dunno. I can't tell if we never run this, and it works anyway, or if we actually do run it despite not being declared `C`. In any case, all other instances of JNI_OnLoad is declared as `C`, so it will certainly not hurt. And, as I said in the comment, when compiling this for static builds it is necessary for the macro to work.

File: src/java.desktop/windows/native/libawt/windows/awt_Toolkit.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (aivanov-jdk) [reply src/java.desktop/windows/native/libawt/windows/awt_Toolkit.cpp:108]: Yep. A static function could be optimised out… inlined. However, if the address of the function is taken, the function should remain as a stand-alone callable function. Adding `static` shouldn't remove the function, it just ensures the function name isn't visible to other .obj files.

I'm not 100% sure, yet I think it should work correctly even when `GetAwtModuleHandle` is declared `static`.

**Review by prrace — APPROVED** [review]

**Comment by prrace** [comment]
> [](https://bugs.openjdk.org/browse/JDK-8346306) is the reason why no destructors are modified anymore, is it?

Much happier that is the case.

File: src/java.desktop/windows/native/libawt/windows/awt_Toolkit.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (magicus) [reply src/java.desktop/windows/native/libawt/windows/awt_Toolkit.cpp]: You are 100% correct. It works fine with static. I got kind of confused that name resolution should come into play, but I guess Windows just looks up the address and determines what module it belongs to that way.

**Comment by magicus** [comment]
@prrace @aivanov-jdk Can you please re-review the latest version (which makes GetAwtModuleHandle static)?

**Review by aivanov-jdk — APPROVED** [review]

**Comment by magicus** [comment]
Thanks!

/integrate
